ipa-core/src/protocol/dp/insecure.rs

```diff
@@ -101,6 +126,98 @@ impl DiscreteDp {
     }
 }
 
+///  Non-negative DP noise for OPRF padding
+///  Samples from a Truncated Double Geometric
+#[derive(Debug, PartialEq)]
+pub struct OPRFPaddingDp {
+    epsilon: f64,
+    delta: f64,
+    sensitivity: u32, // $\Delta$
+    truncated_double_geometric: TruncatedDoubleGeometric,
+}
+fn pow_u32(mut base: f64, mut exp: u32) -> f64 {
+    // To avoid type precision loss, we implemented pow for a u32 exponent
+    // like the algorithm here https://docs.rs/num-traits/0.2.15/src/num_traits/pow.rs.html#189
+    if exp == 0 {
+        return 1.0;
+    }
+
+    while exp & 1 == 0 {
+        base = base * base;
+        exp >>= 1;
+    }
+    if exp == 1 {
+        return base;
+    }
+
+    let mut acc = base;
+    while exp > 1 {
+        exp >>= 1;
+        base = base * base;
+        if exp & 1 == 1 {
+            acc *= base;
+        }
+    }
+    acc
+}
+
+fn right_hand_side(n: u32, big_delta: u32, epsilon: f64) -> f64 {
+    // Computes the right hand side of equation (11) in https://arxiv.org/pdf/2110.08177.pdf
+    let r = E.powf(-epsilon);
+    let a = (1.0 - r) / (1.0 + r - 2.0 * (pow_u32(r, n + 1)));
+    let mut result = 0.0;
+    for k in n - big_delta + 1..=n {
+        result += pow_u32(r, k);
+    }
+    a * result
+}
+fn find_smallest_n(big_delta: u32, epsilon: f64, small_delta: f64) -> u32 {
+    // for a fixed set of DP parameters, finds the smallest n that satisfies equation (11)
+    // of https://arxiv.org/pdf/2110.08177.pdf.  This gives the narrowest TruncatedDoubleGeometric
+    // that will satisify the disired DP parameters.
+    for n in big_delta.. {
```

Comment:
This algorithm seems pretty crude.  If there isn't a simple inversion, could this at use a search?

I'm particularly concerned at this level because you have a lot of nested loops here when you trace this all the way down.